Django 1.8+ support

lucaswiman opened the pull request:
@bruth This pull request is a work in progress to add Django 1.8+ support for `modeltree`.

I wanted to submit this earlier in an uncompleted state to ask what versions of Django you're interested in supporting.

All django versions below 1.8 have been end-of-lifed with the release of Django 1.9. Would you be comfortable dropping Django 1.7 and below from list of supported versions?  Supporting Django 1.4+ seems to introduce a lot of conditional importing logic, and Django 1.8 introduced some pretty substantial changes to how the ORM works.

## Changes so far
- Added 1.8-1.9 to the travis config and setup file.
- Added a tox.ini file to make it easy to run tests across multiple versions of Django. [Happy to remove this if you'd rather not have it clutter up the repo.]
- Fixed the basic import bugs from deprecations in various versions of Django.

## Still TODO
- [x] Fix remaining test failures

Thanks very much for the work you've put in to this library!

naegelyd commented:
@lucaswiman You'll need to sign-off your commits([see `-s` option](https://www.kernel.org/pub/software/scm/git/docs/git-commit.html)) since the testing process checks for DCO consent. Seems like the DCO related docs haven't made it into the `modeltree` contributing docs yet. You can find the relevant documentation on a related repo [here](https://github.com/chop-dbhi/varify/blob/master/CONTRIBUTING.md#contributing). That's why all the tests are failing right now.

naegelyd commented:
> Would you be comfortable dropping Django 1.7 and below from list of supported versions?

FWIW, I'd vote for dropping support for Django <= 1.7 and releasing a new major version of `modeltree` supporting Django 1.8 and 1.9 since 1.9 has now been officially released.

lucaswiman commented:
This is getting closer to ready. I've fixed most of the route computation bugs, and most of the remaining test failures involve `ModelTree.add_joins` method. The root cause is that the API of `Query.join` was updated to take a `django.db.models.sql.datastructures.Join` object.  

## Notes on the API changes
Compare the API below:
- [`Query.join` (1.8.7)](https://github.com/django/django/blame/1.8.7/django/db/models/sql/query.py#L896)
- [`Query.join` (1.7.11)](https://github.com/django/django/blame/1.7.11/django/db/models/sql/query.py#L902)
- [`Join.__init__` (1.8.7)](https://github.com/django/django/blame/1.8.7/django/db/models/sql/datastructures.py#L28), which defines its arguments:
> 
```python
    """
    This class is primarily used in Query.alias_map. All entries in alias_map
    must be Join compatible by providing the following attributes and methods:
        - table_name (string)
        - table_alias (possible alias for the table, can be None)
        - join_type (can be None for those entries that aren't joined from
          anything)
        - parent_alias (which table is this join's parent, can be None similarly
          to join_type)
        - as_sql()
        - relabeled_clone()

    """
    def __init__(self, table_name, parent_alias, table_alias, join_type,
                 join_field, nullable):
```

Taking into account tuple unpacking of `connection` in the pre-1.8 `Query.join`, it has the signature:
> `def join(self, (lhs, table, join_cols), reuse=None, nullable=False, join_field=None)`

I've made a table of proposed changes to help clarify the code changes:

| >=1.8 `Join.__init__` arg | <=1.7 `Query.join` arg | Notes                                                                                                                                                                                                                                 |
|---------------------------|------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `table_name`              | `table`                |                                                                                                                                                                                                                                       |
| `parent_alias`            | `lhs`                  | (I think?) "'lhs' is either an existing table alias or a table name." "lhs.l_id1 = table.r_id1" compared with "(parent_alias) which table is this join's parent, can be None similarly41,to join_type"                                |
| `table_alias`             | blank?                 |                                                                                                                                                                                                                                       |
| `join_type`               |                        | May need to be computed based on the logic here: https://github.com/django/django/blame/1.7.11/django/db/models/sql/query.py#L945                                                                                                     |
| `join_field`              | `join_field`           | I think this is required to be nonempty in the new schema, so the first join appended in `.get_joins()` may now be superfluous. See also `join_cols` below.                                                                           |
| `nullable`                | `nullable`             |                                                                                                                                                                                                                                       |
| N/A                       | `reuse=None`           | Still a kwarg on `Query.join`                                                                                                                                                                                                         |
|                           | `join_cols`            | `'join_cols' is a tuple of tuples containing columns to join on ((l_id1, r_id1), (l_id2, r_id2)).` This seems to be dynamically computed now: https://github.com/django/django/blame/1.8.7/django/db/models/sql/datastructures.py#L57 |

lucaswiman commented:
@bruth, @naegelyd: I've got the full Travis build passing. This is ready for review. I've increased the version to 2.0.0.a1, where the major was updated to reflect dropping support for Django <1.8. This might also be a good time to drop support for deprecated behavior, but that's outside the scope of this PR. IMO, 1.7/1.8 is a good dividing line, since the introduction of a supported `Meta` API changed a lot of the internals of tree building, and allowed simplifying the code.

I think the coveralls failure is caused by removing more lines of code than I added, meaning that previously uncovered lines of code now represent a slightly bigger proportion of the whole. There're a few relevant uncovered lines of code I could add tests for if you want.

Thanks!

lucaswiman commented on modeltree/tree.py:
```diff
@@ -921,10 +850,15 @@ def add_joins(self, model, queryset=None, **kwargs):
+        for i, join in enumerate(self.get_joins(model)):
+            if isinstance(join, BaseTable):
+                alias_map = clone.query.alias_map
+                if join.table_alias in alias_map or \
```
This if statement is something where I don't entirely understand *why* it's required. The old behavior caused a bunch of extraneous table aliases to be constructed in some cases, and just never adding the `BaseTable` joins mean some invalid queries were generated. That said, the tests seem fairly extensive and subtle for `select` construction, so I think the _behavior_ is correct.

naegelyd commented on modeltree/tree.py:
```diff
@@ -104,140 +107,63 @@ def __repr__(self):
-    def get_joins(self, **kwargs):
-        """Returns a list of connections that need to be added to a
+    def get_joins(self):
+        """Returns a list of Join objects that need to be added to a
```
This docstring seems to conflict with the values being returned:

```python
return BaseTable(self.parent.db_table, alias=None), joins
```

Let's update the docstring or just return the joins if the table is not needed in the returned values.

naegelyd commented on modeltree/tree.py:
```diff
@@ -649,14 +575,14 @@ def _filter_m2m(self, field):
 
     def _filter_related_m2m(self, rel):
-        """Tests if this RelatedObject represents a ManyToManyField.
+        """Tests if this relation represents a ManyToManyField.
```
I think this docstring should read `Tests if this relation represents a ManyToManyRel.`

lucaswiman replied:
That doesn't quite work for keeping the docstrings consistent with all the different related types. This section of the code is now pretty repetitive, so I refactored it in ed041c82913b11f9c818e4d869bbbf530730a1be

naegelyd commented on tox.ini:
```diff
@@ -1,3 +1,26 @@
 
 filename = *.py
+
+[tox]
```
Nice. Thanks for adding this all to the tox file.

naegelyd commented:
@lucaswiman These changes look fine to me but I'm going to defer to @bruth to do a final review and merge since he is a lot more familiar with the code here.

bruth commented on modeltree/tree.py:
```diff
@@ -701,106 +564,61 @@ def _add_node(self, parent, model, relation, reverse, related_name,
+            elif f.one_to_many or f.many_to_one:
+                return 'foreignkey'
+            else:
+                raise ValueError(f)  # pragma: no cover
```
Nit. No need to put this in the `else` block since all previous blocks return.

bruth commented:
Thank you for putting in the effort to work on this! The code appears to be good, although I am not familiar with the ORM internals for recent versions of Django, so I will likely not be able to pick out any particular issues. The set of tests are pretty extensive, so we should be able to rely on them to validate the majority of cases.

lucaswiman commented:
What's your release policy? I set this to an alpha version, but I'm not sure what additional testing you'd need done before a full release. I'd also be happy to follow up with a pull request removing deprecations and covering more of `tree.py` with tests before a full release.

naegelyd commented:
> I'd also be happy to follow up with a pull request removing deprecations and covering more of tree.py with tests before a full release.

I'd be fine if this was merged as is at `2.0.0(alpha)` and then the follow up PR adding tests and removing deprecations could set the version to `2.0.0(final)`. @bruth, thoughts?

lucaswiman commented:
Actually hold off on releasing this. In doing integration testing with the code at work, I've run into some bugs with `_find_relations`. I'll try to reproduce the issue with a testcase.

lucaswiman commented:
>  hold off on releasing this

I fixed the issues uncovered by trying them out on our application code. I added two new tests: one for handling of Generic Foreign Keys, and the other for non-proxy model inheritance.

bruth commented:
> I'd be fine if this was merged as is at 2.0.0(alpha) and then the follow up PR adding tests and removing deprecations could set the version to 2.0.0(final).

I agree. Everything appears to be in working order. Do you mind squashing some of the commits are more like amends of other ones? I don't think everything should be squashed down since there are some significant bits, but the minor things could be cleaned up.

lucaswiman commented:
@bruth I rationalized the commit history, into a series of more logically-self-contained commits. Broadly:
- Drop support for older versions of django in the test suite: dd9527f0c820582c624e7d038df9859a53cf759c.
- Make "obvious" changes to remove support for older versions: 0ecb46dcdbd8ba6fb8eab103f04c6d3469379967 (Switch statements on versions, import path changes, etc.)
- A few specific more intricate updates described in 76e7650155e9631cbd34fc704b7501c25f31084e and a74e32077860cd782883ff5b0aeb950ab13c83d4
- Change join behavior: 446493bf5b4a3114d98808dced6e28754820ee4b
- Separate out the one difficult to understand piece of updates to join construction: a7e114d773aee50d520a3b9494f10ca3cbfb85fd
- Big refactor and simplifications to `_find_relations`: 00167557bb52529592bddc38e007d315b8362f4b
- Tests and fix for specific issues discovered from using it in a more complex django application: 
0cefa2a0c36f05eeb209be66c815c74b273e85e7
- Bump version: 96b062e86c1283385f6d08d9355265334846d4b7

I verified that the test suite still passes. Please take a look, and merge if it looks good to you. I'll follow up with the final release changes as described above once this is merged.

Thanks for bearing with me on this!

bruth commented:
Awesome, thanks for all the work.
